redaxo/src/core/lib/sql/table_code_generator.php

```diff
@@ -0,0 +1,169 @@
+    }
+
+    private function getColumn(rex_sql_column $column)
+    {
+        $parameters = [];
+        $nonDefault = false;
+
+        if (null !== $column->getExtra()) {
+            $parameters[] = $this->scalar($column->getExtra());
+            $nonDefault = true;
+        }
+
+        if ($nonDefault || null !== $column->getDefault()) {
+            $parameters[] = $this->scalar($column->getDefault());
+            $nonDefault = true;
+        }
+
+        if ($nonDefault || false !== $column->isNullable()) {
+            $parameters[] = $this->scalar($column->isNullable());
+        }
+
+        $parameters[] = $this->scalar($column->getType());
+        $parameters[] = $this->scalar($column->getName());
+
+        return 'new rex_sql_column('.implode(', ', array_reverse($parameters)).')';
+    }
+
+    private function getIndex(rex_sql_index $index)
+    {
+        $parameters = [
+            $this->scalar($index->getName()),
+            $this->simpleArray($index->getColumns()),
+        ];
+
+        static $types = [
+            rex_sql_index::UNIQUE => 'rex_sql_index::UNIQUE',
+            rex_sql_index::FULLTEXT => 'rex_sql_index::FULLTEXT',
+        ];
+
+        if (rex_sql_index::INDEX !== $index->getType()) {
+            $parameters[] = $types[$index->getType()];
+        }
+
+        return 'new rex_sql_index('.implode(', ', $parameters).')';
+    }
+
+    private function getForeignKey(rex_sql_foreign_key $foreignKey)
+    {
+        $parameters = [
+            $this->scalar($foreignKey->getName()),
+            $this->tableName($foreignKey->getTable()),
+            $this->map($foreignKey->getColumns()),
+        ];
+
+        static $options = [
+            rex_sql_foreign_key::RESTRICT => 'rex_sql_foreign_key::RESTRICT',
+            rex_sql_foreign_key::CASCADE => 'rex_sql_foreign_key::CASCADE',
+            rex_sql_foreign_key::SET_NULL => 'rex_sql_foreign_key::SET_NULL',
+        ];
+
+        $nonDefaultOnDelete = rex_sql_foreign_key::RESTRICT !== $foreignKey->getOnDelete();
+
+        if ($nonDefaultOnDelete || rex_sql_foreign_key::RESTRICT !== $foreignKey->getOnUpdate()) {
+            $parameters[] = $options[$foreignKey->getOnUpdate()];
+        }
+
+        if ($nonDefaultOnDelete) {
+            $parameters[] = $options[$foreignKey->getOnDelete()];
+        }
+
+        return 'new rex_sql_foreign_key('.implode(', ', $parameters).')';
+    }
+
+    private function getPrimaryKey(array $primaryKey)
+    {
+        if (1 === count($primaryKey)) {
+            return $this->scalar(reset($primaryKey));
+        }
+
+        return $this->simpleArray($primaryKey);
+    }
+
+    private function tableName($name)
+    {
+        if (0 !== strpos($name, rex::getTablePrefix())) {
+            return $this->scalar($name);
+        }
+
+        $name = substr($name, strlen(rex::getTablePrefix()));
+
+        return 'rex::getTable('.$this->scalar($name).')';
+    }
+
+    private function scalar($scalar)
```

Comment:
Diese allg. Code generator dinge ggf. In separater klasse?

Reply:
Die sind nicht unbedingt allumfassend, sondern eher speziell für die hier vorkommenden Dinge.
Daher aktuell lieber nicht.